### DISCUSSION TITLE
API for creating webhooks

### DISCUSSION THREAD

**Discussion by filipesmedeiros** [strapi/strapi#6168] [Legacy Topics]
- [ ] **I have created my request on the Product Board before I submitted this issue**
- [X] **I have looked at all the other requests on the Product Board before I submitted this issue**

Hey! Can we have a way to create webhooks programatically? Maybe with a file like what happens with cron tasks, or by making it available through the API?

**Comment by lauriejim** [comment]
Hello, yes it's possible but no documented for now.
Ping @alexandrebodin about that.

I will give you some insight.

Here is the code that load existing WebHooks - [code](https://github.com/strapi/strapi/blob/d1d13fab2393267211a6cb186f18d490a69b640f/packages/strapi/lib/Strapi.js#L334)
We can see this function `this.webhookRunner.add(webhook)` that could be used in your app w/ `strapi.webhookRunner.add(webhook)` let you add a WebHook on the flight.

Here is the structure of the `webhook` object.

```js
{
  name: 'azd',
  url: 'http://toto.fr',
  headers: { From: 'azd' },
  events: [ 'entry.create', 'entry.update', 'entry.delete' ],
  isEnabled: true
}
```

Here you will be able to see the code source of the WebHook API that is used by the admin panel to create/update/delete WebHooks - [code](https://github.com/strapi/strapi/blob/7123f68737cc9a759f7951dba4cd30db0b9fbe8b/packages/strapi-admin/controllers/Webhooks.js#L89)

You will find all possible events in the `ALLOWED_EVENTS` variable.

Then in your use case, you can easily create a file in your config folder with an array of WebHooks definition. And in your `config/bootstrap.json` file, you are using `strapi.config[thePathOfYourConfig]` for get your WebHooks and then use a loop of `strapi.webhookRunner.add(webhook)` to add them.

**Reply by filipesmedeiros** [reply]
Amazing! Thank you!

**Comment by alexandrebodin** [comment]
Be aware this is not a public facing API and this can break on any given release. Until we make it public make sure to verify if an update happened in the internal code if you are using it when upgrading.
